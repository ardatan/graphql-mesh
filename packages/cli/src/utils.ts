export const stripJSONComments = (data: string) => {
  const re = /\/\/(.*)/g;
  const re2 = /\/\*(.*)\*\//g;
  return data.replace(re, '').replace(re2, '');
};
