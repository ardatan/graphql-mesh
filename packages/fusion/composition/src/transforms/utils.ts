export class TransformValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TransformValidationError';
  }
}
