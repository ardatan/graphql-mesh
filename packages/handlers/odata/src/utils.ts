export function getUrlString(url: URL) {
  return decodeURIComponent(url.toString()).split('+').join(' ');
}

export function addIdentifierToUrl(url: URL, identifierFieldName: string, identifierFieldTypeRef: string, args: any) {
  url.href += `/${args[identifierFieldName]}/`;
}
