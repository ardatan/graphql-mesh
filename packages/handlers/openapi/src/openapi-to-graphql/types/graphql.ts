// Copyright IBM Corp. 2018. All Rights Reserved.
// Node module: openapi-to-graphql
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

/**
 * Custom type definitions for GraphQL.
 */

import {
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLInputObjectType,
  GraphQLList,
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLResolveInfo,
} from 'graphql';

export enum GraphQLOperationType {
  Query,
  Mutation,
  // TODO: Subscription
}

export type GraphQLType =
  | GraphQLObjectType
  | GraphQLInputObjectType
  | GraphQLList<any>
  | GraphQLUnionType
  | GraphQLEnumType
  | GraphQLScalarType;

type Arg = {
  type: any;
  description?: string;
};

export type Args = {
  [key: string]: Arg;
};

export type ResolveFunction = (root: any, args: any, ctx: any, info: GraphQLResolveInfo) => Promise<any> | any;

export type Field = {
  type: GraphQLType;
  resolve?: ResolveFunction;
  args?: Args;
  description: string;
};
