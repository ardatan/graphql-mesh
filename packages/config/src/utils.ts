import { parse } from 'graphql';
import { MeshHandlerLibrary, KeyValueCache, YamlConfig, MergerFn, ImportFn, MeshPubSub } from '@graphql-mesh/types';
import { resolve } from 'path';
import { IResolvers, printSchemaWithDirectives } from '@graphql-tools/utils';
import { paramCase } from 'param-case';
import { loadTypedefs } from '@graphql-tools/load';
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader';
import { get, set, kebabCase } from 'lodash';
import { stringInterpolator } from '@graphql-mesh/utils';
import { mergeResolvers } from '@graphql-tools/merge';
import { PubSub, withFilter } from 'graphql-subscriptions';
import { EventEmitter } from 'events';
import { CodeFileLoader } from '@graphql-tools/code-file-loader';
import StitchingMerger from '@graphql-mesh/merger-stitching';

export async function getPackage<T>(name: string, type: string, importFn: ImportFn): Promise<T> {
  const casedName = paramCase(name);
  const casedType = paramCase(type);
  const possibleNames = [
    `@graphql-mesh/${casedName}`,
    `@graphql-mesh/${casedName}-${casedType}`,
    `@graphql-mesh/${casedType}-${casedName}`,
    casedName,
    `${casedName}-${casedType}`,
    `${casedType}-${casedName}`,
    casedType,
  ];
  if (name.includes('-')) {
    possibleNames.push(name);
  }
  const possibleModules = possibleNames.concat(resolve(process.cwd(), name));

  for (const moduleName of possibleModules) {
    try {
      const exported = await importFn(moduleName);

      return (exported.default || exported.parser || exported) as T;
    } catch (err) {
      if (
        !err.message.includes(`Cannot find module '${moduleName}'`) &&
        !err.message.includes(`Could not locate module`)
      ) {
        throw new Error(`Unable to load ${type} matching ${name}: ${err.message}`);
      }
    }
  }

  throw new Error(`Unable to find ${type} matching ${name}`);
}

export async function getHandler(name: keyof YamlConfig.Handler, importFn: ImportFn): Promise<MeshHandlerLibrary> {
  const handlerFn = await getPackage<MeshHandlerLibrary>(name.toString(), 'handler', importFn);

  return handlerFn;
}

export async function resolveAdditionalTypeDefs(baseDir: string, additionalTypeDefs: string) {
  if (additionalTypeDefs) {
    const sources = await loadTypedefs(additionalTypeDefs, {
      cwd: baseDir,
      loaders: [new CodeFileLoader(), new GraphQLFileLoader()],
    });
    return sources.map(source => source.document || parse(source.rawSDL || printSchemaWithDirectives(source.schema)));
  }
  return undefined;
}

export async function resolveAdditionalResolvers(
  baseDir: string,
  additionalResolvers: (
    | string
    | YamlConfig.AdditionalStitchingResolverObject
    | YamlConfig.AdditionalSubscriptionObject
  )[],
  importFn: ImportFn,
  pubsub: MeshPubSub
): Promise<IResolvers> {
  const loadedResolvers = await Promise.all(
    (additionalResolvers || []).map(async additionalResolver => {
      if (typeof additionalResolver === 'string') {
        const filePath = additionalResolver;

        const exported = await importFn(resolve(baseDir, filePath));
        let resolvers = null;

        if (exported.default) {
          if (exported.default.resolvers) {
            resolvers = exported.default.resolvers;
          } else if (typeof exported.default === 'object') {
            resolvers = exported.default;
          }
        } else if (exported.resolvers) {
          resolvers = exported.resolvers;
        }

        if (!resolvers) {
          console.warn(`Unable to load resolvers from file: ${filePath}`);

          return {};
        }

        return resolvers;
      } else {
        if ('pubsubTopic' in additionalResolver) {
          return {
            [additionalResolver.type]: {
              [additionalResolver.field]: {
                subscribe: withFilter(
                  (root, args, context, info) => {
                    const resolverData = { root, args, context, info };
                    const topic = stringInterpolator.parse(additionalResolver.pubsubTopic, resolverData);
                    return pubsub.asyncIterator(topic);
                  },
                  (root, args, context, info) => {
                    return additionalResolver.filterBy ? eval(additionalResolver.filterBy) : true;
                  }
                ),
                resolve: (payload: any) => {
                  if (additionalResolver.returnData) {
                    return get(payload, additionalResolver.returnData);
                  }
                  return payload;
                },
              },
            },
          };
        } else {
          return {
            [additionalResolver.type]: {
              [additionalResolver.field]: {
                selectionSet: additionalResolver.requiredSelectionSet,
                resolve: async (root: any, args: any, context: any, info: any) => {
                  const resolverData = { root, args, context, info };
                  const methodArgs: any = {};
                  for (const argPath in additionalResolver.args) {
                    set(methodArgs, argPath, stringInterpolator.parse(additionalResolver.args[argPath], resolverData));
                  }
                  const result = await context[additionalResolver.targetSource].api[additionalResolver.targetMethod](
                    methodArgs,
                    {
                      selectedFields: additionalResolver.resultSelectedFields,
                      selectionSet: additionalResolver.resultSelectionSet,
                      depth: additionalResolver.resultDepth,
                    }
                  );
                  return additionalResolver.returnData ? get(result, additionalResolver.returnData) : result;
                },
              },
            },
          };
        }
      }
    })
  );

  return mergeResolvers(loadedResolvers);
}

export async function resolveCache(
  cacheConfig: YamlConfig.Config['cache'],
  importFn: ImportFn
): Promise<KeyValueCache | undefined> {
  if (cacheConfig) {
    const cacheName = Object.keys(cacheConfig)[0];
    const config = cacheConfig[cacheName];

    const moduleName = kebabCase(cacheName.toString());
    const pkg = await getPackage<any>(moduleName, 'cache', importFn);
    const Cache = pkg.default || pkg;

    return new Cache(config);
  }
  const InMemoryLRUCache = await import('@graphql-mesh/cache-inmemory-lru').then(m => m.default);
  const cache = new InMemoryLRUCache();
  return cache;
}

export async function resolvePubSub(
  pubsubYamlConfig: YamlConfig.Config['pubsub'],
  importFn: ImportFn
): Promise<MeshPubSub> {
  if (pubsubYamlConfig) {
    let pubsubName: string;
    let pubsubConfig: any;
    if (typeof pubsubYamlConfig === 'string') {
      pubsubName = pubsubYamlConfig;
    } else {
      pubsubName = pubsubYamlConfig.name;
      pubsubConfig = pubsubYamlConfig.config;
    }

    const moduleName = kebabCase(pubsubName.toString());
    const pkg = await getPackage<any>(moduleName, 'pubsub', importFn);
    const PubSub = pkg.default || pkg;

    return new PubSub(pubsubConfig);
  } else {
    const eventEmitter = new EventEmitter({ captureRejections: true });
    eventEmitter.setMaxListeners(Infinity);
    const pubsub = new PubSub({ eventEmitter }) as MeshPubSub;

    return pubsub;
  }
}

export async function resolveMerger(mergerConfig: YamlConfig.Config['merger'], importFn: ImportFn): Promise<MergerFn> {
  if (mergerConfig) {
    const pkg = await getPackage<any>(mergerConfig, 'merger', importFn);
    return pkg.default || pkg;
  }
  return StitchingMerger;
}
