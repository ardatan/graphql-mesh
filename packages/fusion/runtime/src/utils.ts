import { print, type DocumentNode, type ExecutionResult, type GraphQLSchema } from 'graphql';
import type { TransportOptions } from '@graphql-mesh/serve-runtime';
import type {
  Transport,
  TransportContext,
  TransportEntry,
  TransportGetSubgraphExecutor,
  TransportGetSubgraphExecutorOptions,
} from '@graphql-mesh/transport-common';
import type { Logger } from '@graphql-mesh/types';
import { iterateAsync, mapMaybePromise } from '@graphql-mesh/utils';
import {
  isAsyncIterable,
  isDocumentNode,
  mapAsyncIterator,
  printSchemaWithDirectives,
  type ExecutionRequest,
  type Executor,
  type Maybe,
  type MaybePromise,
} from '@graphql-tools/utils';

export type { TransportEntry, TransportGetSubgraphExecutor, TransportGetSubgraphExecutorOptions };

export type Transports =
  | {
      [Kind in string]: MaybePromise<
        | Transport<Kind> // named export
        | {
            default: Transport<Kind>; // property on default export
          }
      >;
    }
  | (<Kind extends string>(
      kind: Kind,
    ) => MaybePromise<
      | Transport<Kind> // named export
      | {
          default: Transport<Kind>; // property on default export
        }
    >);

export type TransportExecutorFactoryGetter = (
  kind: string,
) => MaybePromise<TransportGetSubgraphExecutor>;

function createTransportExecutorFactoryGetter(
  transports: Transports,
  transportOptions: TransportOptions,
  logger: Logger,
): TransportExecutorFactoryGetter {
  return async function getTransport(kind) {
    let transportGetSubgraphExecutor: TransportGetSubgraphExecutor;
    let transport =
      typeof transports === 'function' ? await transports(kind) : await transports?.[kind];
    if (transport) {
      transportGetSubgraphExecutor =
        'default' in transport
          ? transport.default.getSubgraphExecutor
          : transport.getSubgraphExecutor;
    } else {
      const childLogger = logger.child?.(kind);
      try {
        transport = await import(`@graphql-mesh/transport-${kind}`);
      } catch (err) {
        childLogger.error(err);
        throw new Error(
          `No transport found for ${kind}. Please make sure you have installed @graphql-mesh/transport-${kind} or defined the transport config in "mesh.config.ts"`,
        );
      }
      transportGetSubgraphExecutor =
        'default' in transport
          ? transport.default.getSubgraphExecutor
          : transport.getSubgraphExecutor;
      if (!transportGetSubgraphExecutor) {
        throw new Error(
          `@graphql-mesh/transport-${kind} module does not export "getSubgraphExecutor"`,
        );
      }
      if (typeof transportGetSubgraphExecutor !== 'function') {
        throw new Error(
          `@graphql-mesh/transport-${kind} module's export "getSubgraphExecutor" is not a function`,
        );
      }
    }
    return function getSubgraphExecutor(opts) {
      return transportGetSubgraphExecutor({
        ...opts,
        transportEntry: {
          ...opts.transportEntry,
          options: {
            ...opts.transportEntry.options,
            ...transportOptions?.['*']?.[kind],
            ...transportOptions?.[opts.transportEntry.subgraph]?.[kind],
          },
        },
      });
    };
  };
}

function getTransportExecutor(
  transportExecutorFactoryGetter: TransportExecutorFactoryGetter,
  transportContext: TransportGetSubgraphExecutorOptions,
  transportExecutorStack: AsyncDisposableStack,
): MaybePromise<Executor> {
  const kind = transportContext.transportEntry?.kind || '';
  const subgraphName = transportContext.subgraphName || '';
  transportContext.logger?.info(`Loading transport ${kind} for subgraph ${subgraphName}`);
  return mapMaybePromise(transportExecutorFactoryGetter(kind), executor =>
    mapMaybePromise(executor(transportContext), executor => {
      if (isDisposable(executor)) {
        transportExecutorStack.use(executor);
      }
      return executor;
    }),
  );
}

/**
 * This function creates a executor factory that uses the transport packages,
 * and wraps them with the hooks
 */
export function getOnSubgraphExecute({
  onSubgraphExecuteHooks,
  transportContext,
  transportEntryMap,
  getSubgraphSchema,
  transportExecutorStack,
  transports,
  transportOptions,
}: {
  onSubgraphExecuteHooks: OnSubgraphExecuteHook[];
  transports?: Transports;
  transportOptions?: TransportOptions;
  transportContext?: TransportContext;
  transportEntryMap?: Record<string, TransportEntry>;
  getSubgraphSchema(subgraphName: string): GraphQLSchema;
  transportExecutorStack: AsyncDisposableStack;
}) {
  const subgraphExecutorMap = new Map<string, Executor>();
  const transportExecutorFactoryGetter = createTransportExecutorFactoryGetter(
    transports,
    transportOptions,
    transportContext.logger,
  );
  return function onSubgraphExecute(subgraphName: string, executionRequest: ExecutionRequest) {
    let executor: Executor = subgraphExecutorMap.get(subgraphName);
    // If the executor is not initialized yet, initialize it
    if (executor == null) {
      transportContext?.logger?.info(`Initializing executor for subgraph ${subgraphName}`);
      // Lazy executor that loads transport executor on demand
      executor = function lazyExecutor(subgraphExecReq: ExecutionRequest) {
        return mapMaybePromise(
          // Gets the transport executor for the given subgraph
          getTransportExecutor(
            transportExecutorFactoryGetter,
            transportContext
              ? {
                  ...transportContext,
                  subgraphName,
                  get subgraph() {
                    return getSubgraphSchema(subgraphName);
                  },
                  get transportEntry() {
                    return transportEntryMap?.[subgraphName];
                  },
                }
              : {
                  get subgraph() {
                    return getSubgraphSchema(subgraphName);
                  },
                  get transportEntry() {
                    return transportEntryMap?.[subgraphName];
                  },
                  subgraphName,
                },
            transportExecutorStack,
          ),
          executor_ => {
            // Wraps the transport executor with hooks
            executor = wrapExecutorWithHooks({
              executor: executor_,
              onSubgraphExecuteHooks,
              subgraphName,
              transportEntryMap,
              getSubgraphSchema,
            });
            // Caches the executor for future use
            subgraphExecutorMap.set(subgraphName, executor);
            return executor(subgraphExecReq);
          },
        );
      };
      // Caches the lazy executor to prevent race conditions
      subgraphExecutorMap.set(subgraphName, executor);
    }
    return executor(executionRequest);
  };
}

export interface WrapExecuteWithHooksOptions {
  executor: Executor;
  onSubgraphExecuteHooks: OnSubgraphExecuteHook[];
  subgraphName: string;
  transportEntryMap?: Record<string, TransportEntry>;
  getSubgraphSchema: (subgraphName: string) => GraphQLSchema;
}

/**
 * This function wraps the executor created by the transport package
 * with `onSubgraphExecuteHooks` to hook into the execution phase of subgraphs
 */
export function wrapExecutorWithHooks({
  executor,
  onSubgraphExecuteHooks,
  subgraphName,
  transportEntryMap,
  getSubgraphSchema,
}: WrapExecuteWithHooksOptions): Executor {
  if (onSubgraphExecuteHooks.length === 0) {
    return executor;
  }
  return function executorWithHooks(executionRequest: ExecutionRequest) {
    const onSubgraphExecuteDoneHooks: OnSubgraphExecuteDoneHook[] = [];
    return mapMaybePromise(
      iterateAsync(
        onSubgraphExecuteHooks,
        onSubgraphExecuteHook =>
          onSubgraphExecuteHook({
            get subgraph() {
              return getSubgraphSchema(subgraphName);
            },
            subgraphName,
            get transportEntry() {
              return transportEntryMap?.[subgraphName];
            },
            executionRequest,
            setExecutionRequest(newExecutionRequest) {
              executionRequest = newExecutionRequest;
            },
            executor,
            setExecutor(newExecutor) {
              executor = newExecutor;
            },
          }),
        onSubgraphExecuteDoneHooks,
      ),
      () => {
        if (onSubgraphExecuteDoneHooks.length === 0) {
          return executor(executionRequest);
        }
        return mapMaybePromise(executor(executionRequest), currentResult => {
          const executeDoneResults: OnSubgraphExecuteDoneResult[] = [];
          return mapMaybePromise(
            iterateAsync(
              onSubgraphExecuteDoneHooks,
              onSubgraphExecuteDoneHook =>
                onSubgraphExecuteDoneHook({
                  result: currentResult,
                  setResult(newResult: ExecutionResult) {
                    currentResult = newResult;
                  },
                }),
              executeDoneResults,
            ),
            () => {
              if (!isAsyncIterable(currentResult)) {
                return currentResult;
              }

              if (executeDoneResults.length === 0) {
                return currentResult;
              }

              const onNextHooks: OnSubgraphExecuteDoneResultOnNext[] = [];
              const onEndHooks: OnSubgraphExecuteDoneResultOnEnd[] = [];

              for (const executeDoneResult of executeDoneResults) {
                if (executeDoneResult.onNext) {
                  onNextHooks.push(executeDoneResult.onNext);
                }
                if (executeDoneResult.onEnd) {
                  onEndHooks.push(executeDoneResult.onEnd);
                }
              }

              if (onNextHooks.length === 0 && onEndHooks.length === 0) {
                return currentResult;
              }

              const asyncIterator = currentResult[Symbol.asyncIterator]();
              return mapAsyncIterator(
                asyncIterator,
                currentResult =>
                  mapMaybePromise(
                    iterateAsync(onNextHooks, onNext =>
                      onNext({
                        result: currentResult,
                        setResult: res => {
                          currentResult = res;
                        },
                      }),
                    ),
                    () => currentResult,
                  ),
                undefined,
                () =>
                  onEndHooks.length === 0 ? undefined : iterateAsync(onEndHooks, onEnd => onEnd()),
              );
            },
          );
        });
      },
    );
  };
}

export interface UnifiedGraphPlugin {
  onSubgraphExecute?: OnSubgraphExecuteHook;
}

export type OnSubgraphExecuteHook = (
  payload: OnSubgraphExecutePayload,
) => Promise<Maybe<OnSubgraphExecuteDoneHook | void>> | Maybe<OnSubgraphExecuteDoneHook | void>;

export interface OnSubgraphExecutePayload {
  subgraph: GraphQLSchema;
  subgraphName: string;
  transportEntry?: TransportEntry;
  executionRequest: ExecutionRequest;
  setExecutionRequest(executionRequest: ExecutionRequest): void;
  executor: Executor;
  setExecutor(executor: Executor): void;
}

export interface OnSubgraphExecuteDonePayload {
  result: AsyncIterable<ExecutionResult> | ExecutionResult;
  setResult(result: AsyncIterable<ExecutionResult> | ExecutionResult): void;
}

export type OnSubgraphExecuteDoneHook = (
  payload: OnSubgraphExecuteDonePayload,
) => MaybePromise<Maybe<OnSubgraphExecuteDoneResult | void>>;

export type OnSubgraphExecuteDoneResultOnNext = (
  payload: OnSubgraphExecuteDoneOnNextPayload,
) => MaybePromise<void>;

export interface OnSubgraphExecuteDoneOnNextPayload {
  result: ExecutionResult;
  setResult(result: ExecutionResult): void;
}

export type OnSubgraphExecuteDoneResultOnEnd = () => MaybePromise<void>;

export type OnSubgraphExecuteDoneResult = {
  onNext?: OnSubgraphExecuteDoneResultOnNext;
  onEnd?: OnSubgraphExecuteDoneResultOnEnd;
};

export function compareSchemas(
  a: DocumentNode | string | GraphQLSchema,
  b: DocumentNode | string | GraphQLSchema,
) {
  let aStr: string;
  if (typeof a === 'string') {
    aStr = a;
  } else if (isDocumentNode(a)) {
    aStr = print(a);
  } else {
    aStr = printSchemaWithDirectives(a);
  }
  let bStr: string;
  if (typeof b === 'string') {
    bStr = b;
  } else if (isDocumentNode(b)) {
    bStr = print(b);
  } else {
    bStr = printSchemaWithDirectives(b);
  }
  return aStr === bStr;
}

export function isDisposable(obj: any): obj is Disposable | AsyncDisposable {
  return obj?.[Symbol.dispose] != null || obj?.[Symbol.asyncDispose] != null;
}
