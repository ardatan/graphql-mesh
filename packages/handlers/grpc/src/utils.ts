import { KeyValueCache } from '@graphql-mesh/types';
import { readFileOrUrlWithCache } from '@graphql-mesh/utils';
import { ClientReadableStream, ClientUnaryCall, Metadata, MetadataValue } from '@grpc/grpc-js';
import { existsSync } from 'fs';
import { GraphQLEnumTypeConfig } from 'graphql';
import { InputTypeComposer, ObjectTypeComposer, SchemaComposer } from 'graphql-compose';
import { get } from 'lodash';
import { pascalCase } from 'pascal-case';
import { isAbsolute, join } from 'path';
import { IField, Root } from 'protobufjs';

import { getGraphQLScalar, isScalarType } from './scalars';

export type ClientMethod = (
  input: unknown,
  metaData?: Metadata
) => Promise<ClientUnaryCall> | AsyncIterator<ClientReadableStream<unknown>>;

interface InputOutputTypes {
  input: string;
  output: string;
}

export function toSnakeCase(str: string): string {
  return str.split('.').join('_');
}

export function addIncludePathResolver(root: Root, includePaths: string[]): void {
  const originalResolvePath = root.resolvePath;
  root.resolvePath = (origin: string, target: string) => {
    if (isAbsolute(target)) {
      return target;
    }
    for (const directory of includePaths) {
      const fullPath: string = join(directory, target);
      if (existsSync(fullPath)) {
        return fullPath;
      }
    }
    const path = originalResolvePath(origin, target);
    if (path === null) {
      console.warn(`${target} not found in any of the include paths ${includePaths}`);
    }
    return path;
  };
}

export function addMetaDataToCall(
  call: ClientMethod,
  input: unknown,
  context: Record<string, unknown>,
  metaData: Record<string, string | string[] | Buffer>
): Promise<ClientUnaryCall> | AsyncIterator<ClientReadableStream<unknown>> {
  if (metaData) {
    const meta = new Metadata();
    for (const [key, value] of Object.entries(metaData)) {
      let metaValue: unknown = value;
      if (Array.isArray(value)) {
        // Extract data from context
        metaValue = get(context, value);
      }
      // Ensure that the metadata is compatible with what node-grpc expects
      if (typeof metaValue !== 'string' && !(metaValue instanceof Buffer)) {
        metaValue = JSON.stringify(metaValue);
      }

      meta.add(key, metaValue as MetadataValue);
    }

    return call(input, meta);
  }
  return call(input);
}

export async function getBuffer(path: string, cache: KeyValueCache, cwd: string): Promise<Buffer> {
  if (path) {
    const result = await readFileOrUrlWithCache<string>(path, cache, {
      allowUnknownExtensions: true,
      cwd,
    });
    return Buffer.from(result);
  }
  return undefined;
}

export function getTypeName(
  schemaComposer: SchemaComposer<unknown>,
  typePath: string,
  isInput: boolean,
  packageName: string
): string {
  if (isScalarType(typePath)) {
    return getGraphQLScalar(typePath);
  }
  let baseTypeName = pascalCase(typePath);
  const packageNamePrefix = pascalCase(packageName);
  if (baseTypeName.startsWith(packageNamePrefix)) {
    baseTypeName = baseTypeName.replace(packageNamePrefix, '');
  }
  if (isInput && !schemaComposer.isEnumType(baseTypeName)) {
    baseTypeName += 'Input';
  }
  return baseTypeName;
}

export function createEnum(typeName: string, values: Record<string, number>): GraphQLEnumTypeConfig {
  const enumTypeConfig: GraphQLEnumTypeConfig = {
    name: typeName,
    values: {},
  };
  for (const [key, value] of Object.entries(values)) {
    enumTypeConfig.values[key] = {
      value,
    };
  }
  return enumTypeConfig;
}

export function createFieldsType(typeName: string): InputOutputTypes {
  return {
    input: typeName + 'Input',
    output: typeName,
  };
}

export async function addInputOutputFields(
  schemaComposer: SchemaComposer<unknown>,
  inputTC: InputTypeComposer,
  outputTC: ObjectTypeComposer,
  fields: { [k: string]: IField },
  name: string,
  currentPath: string,
  packageName: string
): Promise<void> {
  const fieldKeys = Object.keys(fields);
  if (!fieldKeys.length) {
    // This is a empty proto type
    inputTC.addFields({
      _: {
        type: () => {
          return getTypeName(schemaComposer, 'bool', true, packageName);
        },
      },
    });
    outputTC.addFields({
      _: {
        type: () => {
          return getTypeName(schemaComposer, 'bool', true, packageName);
        },
      },
    });
  }
  await Promise.all(
    fieldKeys.map(async fieldName => {
      const { type, rule } = fields[fieldName];
      let fullType = type;
      if (
        packageName &&
        !isScalarType(type) &&
        !type.includes('.') &&
        currentPath.length &&
        name.includes(currentPath)
      ) {
        fullType = pascalCase(currentPath + type);
      }

      inputTC.addFields({
        [fieldName]: {
          type: () => {
            const inputTypeName = getTypeName(schemaComposer, fullType, true, packageName);
            return rule === 'repeated' ? `[${inputTypeName}]` : inputTypeName;
          },
        },
      });
      outputTC.addFields({
        [fieldName]: {
          type: () => {
            const typeName = getTypeName(schemaComposer, fullType, false, packageName);
            return rule === 'repeated' ? `[${typeName}]` : typeName;
          },
        },
      });
    })
  );
}

export function createInputOutput(
  schemaComposer: SchemaComposer<unknown>,
  name: string,
  currentPath: string,
  packageName: string,
  fields: { [k: string]: IField }
): void {
  const { input, output } = createFieldsType(name);

  const inputTC = schemaComposer.createInputTC({
    name: input,
    fields: {},
  });
  const outputTC = schemaComposer.createObjectTC({
    name: output,
    fields: {},
  });
  addInputOutputFields(
    schemaComposer,
    inputTC,
    outputTC,
    fields,
    name,
    pascalCase(toSnakeCase(currentPath)),
    packageName
  );
}
