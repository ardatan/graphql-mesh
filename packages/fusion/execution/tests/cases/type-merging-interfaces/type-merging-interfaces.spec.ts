import { parse } from 'graphql';
import { composeSubgraphs } from '@graphql-mesh/fusion-composition';
import { getExecutorForFusiongraph } from '@graphql-mesh/fusion-runtime';
import { createDefaultExecutor } from '@graphql-tools/delegate';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import {
  createExecutablePlanForOperation,
  serializeExecutableOperationPlan,
} from '../../../src/operations';
import { productsSchema } from './services/products';
import { storefrontsSchema } from './services/storefronts';

describe('Cross-service interfaces', () => {
  const fusiongraph = composeSubgraphs([
    {
      name: 'products',
      schema: productsSchema,
    },
    {
      name: 'storefronts',
      schema: storefrontsSchema,
    },
  ]);
  const { fusiongraphExecutor } = getExecutorForFusiongraph({
    fusiongraph,
    transports() {
      return {
        getSubgraphExecutor({ subgraphName }) {
          switch (subgraphName) {
            case 'products':
              return createDefaultExecutor(productsSchema);
            case 'storefronts':
              return createDefaultExecutor(storefrontsSchema);
          }
        },
      };
    },
  });
  it('composes the schema', async () => {
    expect(printSchemaWithDirectives(fusiongraph)).toMatchSnapshot();
  });
  const exampleQuery = parse(/* GraphQL */ `
    query {
      storefront(id: "1") {
        id
        name
        productOfferings {
          __typename
          id
          name
          price
          ... on ProductDeal {
            products {
              name
              price
            }
          }
        }
      }
    }
  `);
  it('plans correctly', async () => {
    const plan = createExecutablePlanForOperation({
      fusiongraph,
      document: exampleQuery,
    });
    expect(serializeExecutableOperationPlan(plan)).toMatchSnapshot();
  });
  it('should work', async () => {
    const result = await fusiongraphExecutor({
      document: exampleQuery,
    });
    expect(result).toEqual({
      data: {
        storefront: {
          id: '1',
          name: 'eShoppe',
          productOfferings: [
            {
              __typename: 'Product',
              id: '1',
              name: 'iPhone',
              price: 699.99,
            },
            {
              __typename: 'ProductDeal',
              id: '1',
              name: 'iPhone + Survival Guide',
              price: 679.99,
              products: [
                {
                  name: 'iPhone',
                  price: 699.99,
                },
                {
                  name: 'iOS Survival Guide',
                  price: 24.99,
                },
              ],
            },
            {
              __typename: 'Product',
              id: '2',
              name: 'Apple Watch',
              price: 399.99,
            },
          ],
        },
      },
    });
  });
});
