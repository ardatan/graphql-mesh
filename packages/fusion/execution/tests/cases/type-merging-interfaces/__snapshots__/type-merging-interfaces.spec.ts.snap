// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cross-service interfaces composes the schema 1`] = `
"schema {
  query: Query
}

interface ProductOffering @source(subgraph: "products", name: "ProductOffering") @source(subgraph: "storefronts", name: "ProductOffering") {
  id: ID! @source(subgraph: "products", name: "id", type: "ID!") @source(subgraph: "storefronts", name: "id", type: "ID!")
  name: String! @source(subgraph: "products", name: "name", type: "String!")
  price: Float! @source(subgraph: "products", name: "price", type: "Float!")
}

type Product implements ProductOffering @source(subgraph: "products", name: "Product") @source(subgraph: "storefronts", name: "Product") @resolver(subgraph: "products", operation: "query ProductsByIds($Product_id: [ID!]!) { products(ids: $Product_id) }", kind: "BATCH") @resolver(subgraph: "products", operation: "query ProductById($Product_id: ID!) { product(id: $Product_id) }", kind: "FETCH") @variable(subgraph: "products", name: "Product_id", select: "id") @variable(subgraph: "storefronts", name: "Product_id", select: "id") {
  id: ID! @source(subgraph: "products", name: "id", type: "ID!") @source(subgraph: "storefronts", name: "id", type: "ID!")
  name: String! @source(subgraph: "products", name: "name", type: "String!")
  price: Float! @source(subgraph: "products", name: "price", type: "Float!")
}

type Query {
  products(ids: [ID!]!): [Product]! @resolver(subgraph: "products", operation: "query products($ids: [ID!]!) { products(ids: $ids) }") @source(subgraph: "products", name: "products", type: "[Product]!")
  product(id: ID!): Product @resolver(subgraph: "products", operation: "query product($id: ID!) { product(id: $id) }") @source(subgraph: "products", name: "product", type: "Product")
  storefront(id: ID!): Storefront @resolver(subgraph: "storefronts", operation: "query storefront($id: ID!) { storefront(id: $id) }") @source(subgraph: "storefronts", name: "storefront", type: "Storefront")
}

type ProductDeal implements ProductOffering @source(subgraph: "storefronts", name: "ProductDeal") {
  id: ID! @source(subgraph: "storefronts", name: "id", type: "ID!")
  name: String! @source(subgraph: "storefronts", name: "name", type: "String!")
  price: Float! @source(subgraph: "storefronts", name: "price", type: "Float!")
  products: [Product]! @source(subgraph: "storefronts", name: "products", type: "[Product]!")
}

type Storefront @source(subgraph: "storefronts", name: "Storefront") @resolver(subgraph: "storefronts", operation: "query StorefrontById($Storefront_id: ID!) { storefront(id: $Storefront_id) }", kind: "FETCH") @variable(subgraph: "storefronts", name: "Storefront_id", select: "id") {
  id: ID! @source(subgraph: "storefronts", name: "id", type: "ID!")
  name: String! @source(subgraph: "storefronts", name: "name", type: "String!")
  productOfferings: [ProductOffering]! @source(subgraph: "storefronts", name: "productOfferings", type: "[ProductOffering]!")
}"
`;

exports[`Cross-service interfaces plans correctly 1`] = `
{
  "resolverDependencyFieldMap": {
    "storefront": [
      {
        "id": 0,
        "resolverOperationDocument": "query storefront {
  __export: storefront(id: "1") {
    id
    name
    productOfferings {
      __typename
      id
      __variable_1: id
    }
  }
}",
        "subgraph": "storefronts",
      },
    ],
  },
  "resolverOperationNodes": [],
}
`;
